### ISSUE TITLE
ComponentDefinition: Getting array of used instances

### ISSUE THREAD

**Issue by DanRathbun** [SketchUp/api-issue-tracker#45]
### Background
The APIs have methods / functions for getting counts of instances and **used** instances, but only 1 method / function to get an array of **all** instances (used or not.)

_REF Public Forum thread:_ 
https://forums.sketchup.com/t/issue-with-unused-definitions/61531

### Request

Add Ruby methods  / C functions that implement getting an array of the used (and perhaps unused) instances.

### Example

Here is a Ruby refinement that prototypes a method getting the used instances ...

```ruby
module GetUsedInstances
  refine ::Sketchup::ComponentDefinition do
    def get_used_instances()
      if count_used_instances() == 0 ||
         count_used_instances() == count_instances()
        return instances()
      end
      instances().reject do |inst|
        if inst.parent.is_a?(Sketchup::Model) && inst.parent == model()
          next false
        else # it's a Sketchup::ComponentDefinition
          inst.parent.count_used_instances == 0
        end
      end # reject
    end # method
  end # refine overlay module block
end # refinement library module

using GetUsedInstances

cd = Sketchup.active_model.definitions["Some Component"]
used = cd.get_used_instances
```
... and then to get non-used instance array, one can use array subtract ...
```ruby
not_used = cd.instances - used
```
If array subtract is clunky or difficult under C, perhaps then define both the "used" and "unused" varieties for both languages to keep parity.
~

**Comment by thomthom** [comment]
A "used" instance might appear multiple times. Maybe the output should a `InstancePath`s?

**Comment by DanRathbun** [comment]
> A "used" instance might appear multiple times.

True, (in the model entities) but it is still a single object, and as we are "inside" the defintion list here, an instance should appear ONCE in the result array.

> Maybe the output should _be_  `InstancePath` _object_s?

This is apples and oranges. 

**This "feature" is about filtering unused instance references from used instance references (quickly and easily.)**

-----

Started another issue for the InstancePaths feature & discussion: 
[ComponentInstance: getting all InstancePaths for an instance](https://github.com/SketchUp/api-issue-tracker/issues/46)

**Comment by DanRathbun** [comment]
> @thomthom A "used" instance might appear multiple times. 

As we find in the public thread. Coders are bound to make the incorrect assumption that the result array should have the same number of members as is returned by the **`count_used_insatnces`** method.

The assumption is incorrect for **all** cases, but may be true for some cases.

In the public forum thread, Finfa gives an example of an incorrect case, in which a nested group is the one and only member of the result array (as it should be) and should not be parroted 3 times just because it's parent definition has 3 instances.
https://forums.sketchup.com/t/issue-with-unused-definitions/61531/14

------

So if this is ever implemented, the documentation should dispel this incorrect assumption.

**Comment by Eneroth3** [comment]
This would just be a shorthand for my_definition.instances.map(&:instance_paths), assuming instance_path is the method suggested in #46?

**Comment by Eneroth3** [comment]
I think I'm quite confused about this FR. can you give an example of a practical use case?

**Comment by DanRathbun** [comment]
Refer to the referenced public forum thread please. ( I filed this issue as a favor, in that I know how to file FRs.)

It is quite clear that the objects in the returned arrays, from these two FRs are different.
Having them in separate FR issues allows them to be tracked separately, implemented or denied separately.

**Comment by thomthom** [comment]
Logged as SU-40232
